### PR TITLE
Add ADR for choice of EKS ingress controller

### PR DESCRIPTION
**Pull request by kerin**

### CODE REVIEW

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+Date: 2021-08-12
+
+## Status
+
+Accepted
+
+## Context
+
+We require a method of managing and direcing external internet traffic into the cluster. Kubernetes provides [several options for handling inbound traffic](https://medium.com/google-cloud/kubernetes-nodeport-vs-loadbalancer-vs-ingress-when-should-i-use-what-922f010849e0).
```
Comment (fredericfran-gds) [inline docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:11]: Nice that we got the same reference here when we were trying to explore the different options for getting external traffic inside the cluster.

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+We do not want to expose the cluster directly to the internet, and so require an AWS load balancer in front of the cluster. The load balancer must support TLS termination and integration with our DNS provider (AWS Route 53).
+
+Historically Kubernetes has supported provisioning of ALBs and NLBs for `Service` resources of `type=LoadBalancer` via the in-tree (built-in) [AWS cloud provider](https://github.com/kubernetes/cloud-provider-aws), with out-of-tree controllers required for `Ingress` resources. Built-in cloud providers are now [considered deprecated overall, in favour of out-of-tree providers](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/), so an [Ingress Controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) with support for `Service` resources is required.
```
Comment (fredericfran-gds) [inline docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:15]: Thank you for adding this reference here, this backs up the decision nicely.

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+We must also consider how Kubernetes edge services and AWS load balancers will interact with the existing [GOV.UK Router service](https://github.com/alphagov/router), as there is significant overlap in their functionality and responsibilities. This will require further investigation and likely experimentation, and so that end we should ensure that we're able to use both `Ingress` and `Service` Kubernetes resources so that we have the flexibility to support a wide range of use cases in the immediate term - L4 & L7 traffic, name-based routing, HTTP->HTTPS redirection, etc.
+
+## Decision
+
+Use the [AWS Load Balancer Controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller).
+
+## Consequences
+
+The AWS Load Balancer Controller supports TLS certificates via AWS Certificate Manager only, so certificates must be managed there (to be covered in a future ADR).
```
Comment (fredericfran-gds) [inline docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:27]: That's good for our usage as we want all certs used in our AWS infrastructure to auto-renews itself (i.e. self manage) which usually means certs issued by AWS ACM.

**Review by fredericfran-gds — APPROVED** [review]
looks good to me.

I particularly appreciated the good use of external references to support this ADR.

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md (hunk 1 above)
Comment (sengi) [inline docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:11]: typo
```suggestion
We require a method of managing and directing external internet traffic into the cluster. Kubernetes provides [several options for handling inbound traffic](https://medium.com/google-cloud/kubernetes-nodeport-vs-loadbalancer-vs-ingress-when-should-i-use-what-922f010849e0).
```

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Status
+
+Accepted
+
+## Context
+
+We require a method of managing and direcing external internet traffic into the cluster. Kubernetes provides [several options for handling inbound traffic](https://medium.com/google-cloud/kubernetes-nodeport-vs-loadbalancer-vs-ingress-when-should-i-use-what-922f010849e0).
+
+We do not want to expose the cluster directly to the internet, and so require an AWS load balancer in front of the cluster. The load balancer must support TLS termination and integration with our DNS provider (AWS Route 53).
```
Comment (sengi) [inline docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:13]: I see you've already mentioned WAF and Shield below. We're actually using both already in production, so could potentially also mention them here as another reason for wanting an AWS load balancer in front of the cluster.

**Review by sengi — APPROVED** [review]
Cool, all sounds good to me 👍 Thanks for writing up!

File: docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md (hunk 4 above)
Comment (kerin) [reply docs/architecture/decisions/0004-use-aws-load-balancer-controller-for-edge-traffic-services.md:13]: Thanks, added a note on current WAF/Shield usage.

**Pull request merged**